test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/DistributedCacheTagHelperTest.cs

```diff
@@ -0,0 +1,926 @@
+        [InlineData("X-CustomHeader,Accept-Encoding, NotAvailable",
+         "DistributedCacheTagHelper||some-name||VaryByHeader(X-CustomHeader||Header-Value||Accept-Encoding||utf8||NotAvailable||)")]
+        [InlineData("X-CustomHeader,  , Accept-Encoding, NotAvailable",
+         "DistributedCacheTagHelper||some-name||VaryByHeader(X-CustomHeader||Header-Value||Accept-Encoding||utf8||NotAvailable||)")]
+        public void GenerateKey_UsesVaryByHeader(string varyByHeader, string expected)
+        {
+            // Arrange
+            var name = "some-name";
+            var tagHelperContext = GetTagHelperContext();
+            var memoryCache = new MemoryCache(new MemoryCacheOptions());
+            var cacheTagHelper = new DistributedCacheTagHelper(
+                memoryCache,
+                Mock.Of<IDistributedCacheTagHelperStorage>(), 
+                Mock.Of<IDistributedCacheTagHelperFormatter>(), 
+                new HtmlTestEncoder())
+            {
+                ViewContext = GetViewContext(),
+                VaryByHeader = varyByHeader,
+                Name = name
+            };
+            var headers = cacheTagHelper.ViewContext.HttpContext.Request.Headers;
+            headers["Accept-Language"] = "en-us;charset=utf8";
+            headers["Accept-Encoding"] = "utf8";
+            headers["X-CustomHeader"] = "Header-Value";
+
+            // Act
+            var key = cacheTagHelper.GenerateKey(tagHelperContext);
+
+            // Assert
+            Assert.Equal(GetHashedBytes(expected), key);
+        }
+
+        [Theory]
+        [InlineData("category", "DistributedCacheTagHelper||some-name||VaryByQuery(category||cats)")]
+        [InlineData("Category,SortOrder,SortOption",
+            "DistributedCacheTagHelper||some-name||VaryByQuery(Category||cats||SortOrder||||SortOption||Adorability)")]
+        [InlineData("Category,  SortOrder, SortOption,  ",
+            "DistributedCacheTagHelper||some-name||VaryByQuery(Category||cats||SortOrder||||SortOption||Adorability)")]
+        public void GenerateKey_UsesVaryByQuery(string varyByQuery, string expected)
+        {
+            // Arrange
+            var name = "some-name";
+            var tagHelperContext = GetTagHelperContext();
+            var memoryCache = new MemoryCache(new MemoryCacheOptions());
+            var cacheTagHelper = new DistributedCacheTagHelper(
+                memoryCache,
+                Mock.Of<IDistributedCacheTagHelperStorage>(), 
+                Mock.Of<IDistributedCacheTagHelperFormatter>(), 
+                new HtmlTestEncoder())
+            {
+                ViewContext = GetViewContext(),
+                VaryByQuery = varyByQuery,
+                Name = name
+            };
+            cacheTagHelper.ViewContext.HttpContext.Request.QueryString =
+                new Http.QueryString("?sortoption=Adorability&Category=cats&sortOrder=");
```

Comment:
Does `QueryString` conflict with other types? If not, move the `Http.` to a using.